### PR TITLE
ambient upstream: CNI for ambient redirection

### PR DESCRIPTION
**Pull request by howardjohn**
For https://github.com/istio/istio/issues/40879

This PR upstreams the CNI logic for ambient redirection to ztunnel. This is intended to be *off by default* and have no impact for existing users. Both the CNI daemon and CNI plugin have an explicit flag for "ambient enabled" and should perform no new actions when that is unset. The default is *not enabled*.

Quoting from the issue:
> **Reviewer guide**
> 
> It is suggested that reviewers treat these PRs differently than normal, as the code has already been reviewed. PRs should be thoroughly reviewed for risk to non-ambient users, ensuring any new/risky/expensive code is disabled by default. However, code quality comments, nit picks, or even bugs in ambient code should not block a merge; modifying these PRs will be high cost and lead to a 3-way drift from master, experimental-ambient, and the PR. Instead, these comments should be turned into issues to complete following the merge.
> 
> Release notes will not be added for any PR. Instead, they will be written holistically for the ambient feature in 1.18 at a later point.

### CODE REVIEW

File: manifests/charts/istio-cni/values.yaml [hunk 1]
```diff
@@ -44,6 +44,12 @@ cni:
   # Allow the istio-cni container to run in privileged mode, needed for some platforms (e.g. OpenShift)
   privileged: false
 
+  # Configure ambient settings
+  ambient:
+    # If enabled, ambient redirection will be enabled
+    enabled: false
```
Comment (costinm) [inline manifests/charts/istio-cni/values.yaml:50]: Same as the other review - eventually this should move to a configmap outside of values.yaml/install realm.

File: pkg/config/constants/constants.go [hunk 2]
```diff
@@ -150,4 +150,16 @@ const (
...
+
+	// DataplaneMode namespace label for determining ambient mesh behavior
+	DataplaneMode        = "istio.io/dataplane-mode"
+	DataplaneModeAmbient = "ambient"
+
+	// AmbientRedirection specifies whether a pod has ambient redirection (to ztunnel) configured.
+	AmbientRedirection = "ambient.istio.io/redirection"
+	// AmbientRedirectionEnabled indicates redirection is configured. This is set by the CNI when it
+	// actually sets up redirection, rather than by the user.
+	AmbientRedirectionEnabled = "enabled"
```
Comment (costinm) [inline pkg/config/constants/constants.go:162]: Is this what we use for sidecar ( vs true / false ), forgot.

File: pkg/config/constants/constants.go [hunk 3]
```diff
@@ -150,4 +150,16 @@ const (
 	// CertProviderNone does not create any certificates for the control plane. It is assumed that some external
 	// load balancer, such as an Istio Gateway, is terminating the TLS.
 	CertProviderNone = "none"
+
+	// DataplaneMode namespace label for determining ambient mesh behavior
+	DataplaneMode        = "istio.io/dataplane-mode"
+	DataplaneModeAmbient = "ambient"
+
+	// AmbientRedirection specifies whether a pod has ambient redirection (to ztunnel) configured.
+	AmbientRedirection = "ambient.istio.io/redirection"
```
Comment (costinm) [inline pkg/config/constants/constants.go:159]: I wonder if we need 'ambient.istio.io' vs just 'istio.io'. 

Also if 'istio.io/capture=ambientv1' isn't a better model ( with a backfill to add 'sidecarv1' ).

Comment (costinm) [reply pkg/config/constants/constants.go:159]: That's also consistent with DataplaneMode.

File: pkg/config/constants/constants.go [hunk 4]
```diff
@@ -150,4 +150,16 @@ const (
 	// CertProviderNone does not create any certificates for the control plane. It is assumed that some external
 	// load balancer, such as an Istio Gateway, is terminating the TLS.
 	CertProviderNone = "none"
+
+	// DataplaneMode namespace label for determining ambient mesh behavior
+	DataplaneMode        = "istio.io/dataplane-mode"
```
Comment (costinm) [inline pkg/config/constants/constants.go:155]: Do we need -mode ?

**Review by costinm — APPROVED** [review]
I don't think it needs to be off by default - it is already opt-in via namespace label. Not sure why we would need double off.

**Comment by howardjohn** [comment]
> I don't think it needs to be off by default - it is already opt-in via namespace label. Not sure why we would need double off.

It does introduce hostNetwork=true, some hostPath mounts, etc which is why I leaned towards this to reduce risk. It may be something we decide to turn on by default for 1.18 anyways though since those are still relatively low risk

**Pull request merged**